tls/s2n_ktls_parameters.h

```diff
@@ -24,21 +24,66 @@
  * - https://elixir.bootlin.com/linux/v6.3.8/A/ident/SOL_TCP
  */
 
-#define S2N_TLS_ULP_NAME      "tls"
-#define S2N_TLS_ULP_NAME_SIZE sizeof(S2N_TLS_ULP_NAME)
-
 #if defined(__linux__)
     #define S2N_KTLS_SUPPORTED true
 
     /* socket definitions */
     #define S2N_TCP_ULP 31 /* Attach a ULP to a TCP connection.  */
     #define S2N_SOL_TCP 6  /* TCP level */
+    #define S2N_SOL_TLS 282
+    #define S2N_TLS_TX  1 /* Set transmit parameters */
+    #define S2N_TLS_RX  2 /* Set receive parameters */
+
+    /* cipher definitions */
+    #define S2N_TLS_CIPHER_AES_GCM_128              51
+    #define S2N_TLS_CIPHER_AES_GCM_128_IV_SIZE      8
+    #define S2N_TLS_CIPHER_AES_GCM_128_KEY_SIZE     16
+    #define S2N_TLS_CIPHER_AES_GCM_128_SALT_SIZE    4
+    #define S2N_TLS_CIPHER_AES_GCM_128_TAG_SIZE     16
+    #define S2N_TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE 8
 
 #else
-    /* For unsupported platforms 0-init all values. */
-    #define S2N_KTLS_SUPPORTED false
+    /* For unsupported platforms 0-init (array of size 1) all values. */
+    #define S2N_KTLS_SUPPORTED                      false
 
     /* socket definitions */
-    #define S2N_TCP_ULP        0
-    #define S2N_SOL_TCP        0
+    #define S2N_TCP_ULP                             0
+    #define S2N_SOL_TCP                             0
+    #define S2N_SOL_TLS                             0
+    #define S2N_TLS_TX                              0
+    #define S2N_TLS_RX                              0
+
+    /* cipher definitions */
+    #define S2N_TLS_CIPHER_AES_GCM_128              0
+    #define S2N_TLS_CIPHER_AES_GCM_128_IV_SIZE      1
+    #define S2N_TLS_CIPHER_AES_GCM_128_KEY_SIZE     1
+    #define S2N_TLS_CIPHER_AES_GCM_128_SALT_SIZE    1
+    #define S2N_TLS_CIPHER_AES_GCM_128_TAG_SIZE     1
+    #define S2N_TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE 1
+
 #endif
+
+/* Common */
+#define S2N_TLS_ULP_NAME      "tls"
+#define S2N_TLS_ULP_NAME_SIZE sizeof(S2N_TLS_ULP_NAME)
+
+typedef unsigned short __u16;
+
+struct s2n_tls_crypto_info {
+    __u16 version;
+    __u16 cipher_type;
+};
+
+struct s2n_tls12_crypto_info_aes_gcm_128 {
+    struct s2n_tls_crypto_info info;
+    unsigned char iv[S2N_TLS_CIPHER_AES_GCM_128_IV_SIZE];
+    unsigned char key[S2N_TLS_CIPHER_AES_GCM_128_KEY_SIZE];
+    unsigned char salt[S2N_TLS_CIPHER_AES_GCM_128_SALT_SIZE];
+    unsigned char rec_seq[S2N_TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE];
+};
```

Comment:
If you're passing this as-is to setsockopt, I'm skeptical that this is the real definition. I assume the real definition supports ciphers other than AES128? Why are you using this modified definition?

Reply:
https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/tls.h#L125 Oh no it is the real definition. Are they initially casting the void* pointer to s2n_tls_crypto_info in order to get the type, and then casting to the right type? :/

Reply:
https://elixir.bootlin.com/linux/latest/source/net/tls/tls_main.c#L397 some type/len checks [container_of](https://elixir.bootlin.com/linux/latest/source/include/linux/container_of.h#L18).

But as stated above, I am going to get the struct definitions from the linux headers